File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/FragmentLoader.java
package io.github.notenoughupdates.moulconfig.xml.loaders;

import io.github.notenoughupdates.moulconfig.common.MyResourceLocation;
import io.github.notenoughupdates.moulconfig.gui.GuiComponent;
import io.github.notenoughupdates.moulconfig.observer.GetSetter;
import io.github.notenoughupdates.moulconfig.xml.ChildCount;
import io.github.notenoughupdates.moulconfig.xml.XMLContext;
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader;
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse;
import org.w3c.dom.Element;

import javax.xml.namespace.QName;
import java.util.LinkedHashMap;
import java.util.Map;

public class FragmentLoader implements XMLGuiLoader.Basic<GuiComponent> {
    @Override
    public GuiComponent createInstance(XMLContext<?> context, Element element) {
        MyResourceLocation location = context.getPropertyFromAttribute(element, new QName("value"), MyResourceLocation.class).get();
        GetSetter<Object> bind = context.getPropertyFromAttribute(element, new QName("bind"), Object.class);
        return context.getUniverse().load(bind != null ? bind.get() : element, location);
    }
    @Override public QName getName() { return XMLUniverse.qName("Fragment"); }
    @Override public ChildCount getChildCount() { return ChildCount.NONE; }
    @Override public Map<String, Boolean> getAttributeNames() {
        Map<String, Boolean> map = new LinkedHashMap<>();
        map.put("value", true);
        map.put("bind", false);
        return map;
    }
}

File: common/src/main/java/io/github/notenoughupdates/moulconfig/xml/loaders/FragmentLoader.kt
package io.github.notenoughupdates.moulconfig.xml.loaders

import io.github.notenoughupdates.moulconfig.common.MyResourceLocation
import io.github.notenoughupdates.moulconfig.gui.GuiComponent
import io.github.notenoughupdates.moulconfig.xml.ChildCount
import io.github.notenoughupdates.moulconfig.xml.XMLContext
import io.github.notenoughupdates.moulconfig.xml.XMLGuiLoader
import io.github.notenoughupdates.moulconfig.xml.XMLUniverse
import org.w3c.dom.Element
import javax.xml.namespace.QName

class FragmentLoader : XMLGuiLoader.Basic<GuiComponent> {
    override fun createInstance(context: XMLContext<*>, element: Element): GuiComponent {
        val location = context.getPropertyFromAttribute(element, QName("value"), MyResourceLocation::class.java)!!.get()
        val bind = context.getPropertyFromAttribute(element, QName("bind"), Any::class.java)?.get() ?: element
        return context.universe.load(bind, location)
    }

    override fun getName(): QName {
        return XMLUniverse.qName("Fragment")
    }

    override fun getChildCount(): ChildCount {
        return ChildCount.NONE
    }

    override fun getAttributeNames(): Map<String, Boolean> {
        return mapOf("value" to true, "bind" to false)
    }
}
